rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,802 @@
+        node_id: u64,
+        rx: &[u8],
+        tx: &mut [u8],
+        exchange: &mut Exchange<'_>,
+    ) -> Result<bool, Error>
+    where
+        T: DataModelHandler,
+    {
+        let mut wb = WriteBuf::new(tx);
+
+        let req = SubscribeReq::from_tlv(&get_root_node_struct(rx)?)?;
+        let req = ReportDataReq::Subscribe(&req);
+
+        let metadata = self.handler.lock().await;
 
-                    driver.complete(req).await?;
+        let accessor = exchange.accessor()?;
+
+        {
+            let node = metadata.node();
+            let mut attrs = node.read(&req, None, &accessor).peekable();
+
+            loop {
+                let more_chunks = req
+                    .respond(&self.handler, Some(id), &mut attrs, &mut wb, false)
+                    .await?;
+
+                exchange.send(OpCode::ReportData, wb.as_slice()).await?;
+
+                if !Self::recv_status(exchange).await? {
+                    info!("Subscription {node_id:x}::{id} removed during reporting");
+                    return Ok(false);
                 }
-                Interaction::Subscribe {
-                    req,
-                    ref mut driver,
-                } => {
-                    let accessor = driver.accessor()?;
-
-                    'outer: for item in metadata.node().subscribing_read(req, None, &accessor) {
-                        while !AttrDataEncoder::handle_read(&item, &self.0, &mut driver.writer()?)
-                            .await?
-                        {
-                            if !driver.send_chunk(req).await? {
-                                break 'outer;
-                            }
-                        }
-                    }
 
-                    driver.complete(req).await?;
+                if !more_chunks {
+                    break;
                 }
             }
         }
 
+        Ok(true)
+    }
+
+    async fn rx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        let buffer = self.tx_buffer(exchange).await?;
+
+        if let Some(mut buffer) = buffer {
+            let rx = exchange.rx()?;
+
+            buffer.clear();
+            buffer.extend_from_slice(rx.payload()).unwrap();
+
+            exchange.rx_done()?;
+
+            Ok(Some(buffer))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn tx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        if let Some(mut buffer) = self.buffers.get().await {
+            buffer.resize_default(MAX_EXCHANGE_TX_BUF_SIZE).unwrap();
+
+            Ok(Some(buffer))
+        } else {
+            Self::send_status(exchange, IMStatusCode::Busy).await?;
+
+            Ok(None)
+        }
+    }
+
+    async fn recv_status(exchange: &mut Exchange<'_>) -> Result<bool, Error> {
+        let rx = exchange.recv().await?;
+        let opcode = rx.meta().proto_opcode;
+
+        if opcode == OpCode::StatusResponse as u8 {
```

Comment:
I believe this part could benefit from early-return as the if-conditions for return values are complex, even if this is done just for the opcode check. Also wondering if a match would make the usage of resp.status only clearer.

```rust
if opcode != OpCode::StatusResponse as u8 {
    return Err(ErrorCode::Invalid.into());
}

match StatusResp::from_tlv(&get_root_node_struct(rx.payload())?)?.status {
    IMStatusCode::Success => Ok(ReceivedStatus::Success),
    code => {
        warn!("Got status response {:?}, aborting interaction", code);
        drop(rx);
        exchange.acknowledge().await?;
        Ok(ReceivedStatus::Error(code))
    }
}
```

Reply:
Early return I did (though the function body in that particular case is relatively short). `match` I don't think helps that much, as you are matching on one condition only. Which is essentially an `if ... {} else {}` statement. I feel that if I bring your `match` argument to the extreme, we kind of don't even need `if` statements, don't we?

For me, a `match` makes more sense for cases where you have more than one branch. A bit like `switch` in these other languages.